Update the import doc, adding detail about where to find the type token (#3992)
pulumi/pulumi-hugo: themes/default/content/docs/using-pulumi/adopting-pulumi/import/_index.md

- /docs/guides/adopting/import/
---

Most infrastructure as code projects require working with existing cloud resources, whether those resources were originally created with another IaC tool or manually provisioned with a cloud provider console or CLI. Interacting with a previously created cloud resource with Pulumi typically happens in one of two ways:

> * Referencing the properties of the existing cloud resource in order to use those properties to configure a Pulumi-managed resource.

Reviewer: Debatable, and nit-picky, but because it says "one of two ways", I would use an ordered list. Alternatively, could word the preview paragraph as "one of the following ways".

```suggestion
1. Referencing the properties of the existing cloud resource in order to use those properties to configure a Pulumi-managed resource.
```

> * Adopting the existing resource to bring it under management by Pulumi.

Reviewer: ```suggestion
1. Adopting the existing resource to bring it under management by Pulumi.
```

### Where to find the type token and lookup property {#where-to-find}

You'll find the type token and lookup property in the Import section of the resource's API documentation in the [Pulumi Registry](/registry/). The type token is quoted in the `pulumi import` example, and the lookup property can be found in the description just above it:

![Where to find the type token and lookup property for a resource](./token-and-lookup.png)

{{< notes >}}
Make sure the resource provider is configured in a way that allows it to locate the resource you want to import --- e.g., that the resource is in the same region as other resources in your stack --- or the import operation may fail.
{{< /notes >}}

## Importing resources with the CLI

> The `pulumi import` command looks up the resource using the specified type token and resource identifier, adds the resource to the stack's current state, and emits the code required to manage the resource with Pulumi from that point forward. This option requires the least manual effort, so is generally recommended, and is best suited to projects consisting consisting of only one stack.

Reviewer: I think we need to explain what a type token is. I prefer "type identifier" as it's less jargon-y (if it won't make our docs wildly inconsistent) and may not require much more than an example, e.g. `aws:ec2/vpc:Vpc`, to make it clear what it is (a string that tells Pulumi what resource type it is).

Author: Ah yeah, that's better. I'll do that.

Author: It looks like we refer to this as "type token" in more places than I realized. Maybe best to keep to that for consistency for now, but it's something worth discussing for sure.

To import an existing cloud resource with the Pulumi CLI, use the following syntax:

```bash
$ pulumi import <type> <name> <id>
```

* The first argument, `type`, is the Pulumi type token to use for the imported resource.

>     As [mentioned above](#how-import-works), you'll find the type token for a given resource by navigating to the Import section of the resource's API documentation in the [Pulumi Registry](/registry/). The type token of an [Amazon S3 Bucket](/registry/packages/aws/api-docs/s3/bucket/#import) resource, for example, is `aws:s3/bucket:Bucket`.

Reviewer: ```suggestion
    As mentioned in [Where to find the type token and lookup property](#how-import-works), you'll find the type token for a given resource by navigating to the Import section of the resource's API documentation in the [Pulumi Registry](/registry/). For example, the type token of an [Amazon S3 Bucket](/registry/packages/aws/api-docs/s3/bucket/#import) resource, for example, is `aws:s3/bucket:Bucket`.
```

> * The second argument, `name`, is the [resource name](/docs/concepts/resources/names) to apply to the resource once it's imported. The generated code will use this name for the resource declaration, so like all Pulumi resource names, it must be unique within the scope of the containing project.

Reviewer: ```suggestion
* The second argument, `name`, is the [resource name](/docs/concepts/resources/names) to apply to the resource once it's imported. The generated code will use this name for the resource declaration (the first parameter in any resource), so like all Pulumi resource names, it must be unique among all resources for this type within the scope of the containing project. (That is, you may have an S3 bucket and a VPC named `foo`, but you can not have 2 S3 buckets named `foo`.)
```

 =   └─ aws:s3:Bucket     my-bucket   import

Resources:
    + 1 to create
    = 1 to import
    2 changes

Do you want to perform this import?
> yes
  no
  details
```

Choosing to proceed immediately adds the resource to the current stack's state and emits a block of code to `STDOUT` to be added to the Pulumi program. If the current program were written in TypeScript, for example, the resulting CLI output would resemble the following:

```plain
...

Importing (dev)

     Type                 Name        Status
 +   pulumi:pulumi:Stack  dev         created (3s)
 =   └─ aws:s3:Bucket     my-bucket   imported (1s)

Resources:
    + 1 created
    = 1 imported
    2 changes

Duration: 4s

Please copy the following code into your Pulumi application. Not doing so
will cause Pulumi to report that an update will happen on the next update command.

Please note that the imported resources are marked as protected. To destroy them
you will need to remove the `protect` option and run `pulumi update` *before*
the destroy will take effect.

import * as pulumi from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

const my_bucket = new aws.s3.Bucket("infra-logs", {

>     arn: "arn:aws:s3:::company-infra-logs",

Reviewer: It really returns the ARN? I thought that was strictly an output.

Author: It does! This is the actual output, even on a brand-new empty bucket. I actually wonder whether it's worth filing an issue on this, as you can remove everything except `bucket`, run `up`, and get no changes. What do you think?
